src/jvm/main/org/jetbrains/kotlinx/lincheck/Lincheck.kt

```diff
@@ -0,0 +1,87 @@
+import org.jetbrains.kotlinx.lincheck.execution.parallelResults
+import org.jetbrains.kotlinx.lincheck.strategy.LincheckFailure
+import org.jetbrains.kotlinx.lincheck.strategy.managed.ManagedCTestConfiguration
+import org.jetbrains.kotlinx.lincheck.strategy.managed.ManagedStrategy
+import org.jetbrains.kotlinx.lincheck.strategy.managed.forClasses
+import org.jetbrains.kotlinx.lincheck.strategy.managed.modelchecking.ModelCheckingOptions
+import org.jetbrains.kotlinx.lincheck.strategy.verify
+import org.jetbrains.kotlinx.lincheck.transformation.LincheckJavaAgent.ensureObjectIsTransformed
+import org.jetbrains.kotlinx.lincheck.transformation.withLincheckJavaAgent
+import org.jetbrains.kotlinx.lincheck.verifier.Verifier
+
+
+/**
+ * This method will explore different interleavings of the [block] body and all the threads created within it,
+ * searching for the first raised exception.
+ *
+ * @param invocations number of different interleavings of code in the [block] that should be explored.
+ * @param block lambda which body will be a target for the interleavings exploration.
+ */
+fun <R> runConcurrentTest(
+    invocations: Int = DEFAULT_INVOCATIONS_COUNT,
+    block: () -> R
+) {
+    val scenario = scenario {
+        parallel {
+            thread { actor(GeneralPurposeMCWrapper<R>::run, block) }
+        }
+    }
+
+    val options = ModelCheckingOptions()
+        .iterations(0)
+        .invocationsPerIteration(invocations)
+        .addCustomScenario(scenario)
+        .addGuarantee(forClasses(GeneralPurposeMCWrapper::class).allMethods().ignore())
+        .verifier(ExecutionExceptionsVerifier::class.java)
+
+    val testCfg = options.createTestConfigurations(GeneralPurposeMCWrapper::class.java)
+
+    withLincheckJavaAgent(testCfg.instrumentationMode) {
+        ensureObjectIsTransformed(block)
+        val strategy = testCfg.createStrategy(GeneralPurposeMCWrapper::class.java, scenario, null, null)
+        val verifier = testCfg.createVerifier()
+
+        for (i in 1..invocations) {
+            if (!strategy.nextInvocation()) {
+                break
+            }
+            val result = strategy.runInvocation()
+            val failure = strategy.verify(result, verifier)
+            if (failure != null) {
+                throw LincheckAssertionError(failure)
+            }
+        }
+    }
+}
+
+internal class GeneralPurposeMCWrapper<R>() {
+    fun run(block: () -> R) = block()
+}
+
+/**
+ * [ExecutionExceptionsVerifier] makes sure that in case if user `check(...)`'s
+ * fail and store their exceptions as an execution result (or just any exception is thrown),
+ * such outcomes will be disallowed and the execution will fail.
+ */
+private class ExecutionExceptionsVerifier(@Suppress("UNUSED_PARAMETER") sequentialSpecification: Class<*>) : Verifier {
+    override fun verifyResults(scenario: ExecutionScenario?, results: ExecutionResult?): Boolean {
+        if (results == null) return true
```

Comment:
How can `scenario` and `results` be `null`?

Reply:
Fixed. That was just because of java interop nullability issues.